session_id 부여 로직 수정 및 Airflow dag 개발

dodorip opened the pull request:
#4 아래와 같은 이유로 session_id 부여 로직 수정하였습니다.
- 기존 user_id와 session_number 만으로 session_id를 부여할 경우 전체 데이터에선 일관성이 보장되지 않을 가능성 존재

## 로직 수정 사항
1. 동일 세션 내의 event_time 최소값 추출 
 - 컬럼명 : session_start_time
2. session_id 컬럼 로직 변경
- 기존 : user_id & session_number
- 현재 : session_start_time & user_id & session_number
- 최종 : session_start_time & user_id

> session_number가 꼭 필요한가요?
-> session_start_time & user_id -> 세션을 식별하는 식별자로서 session_start_time이 그 역할을 하죠

3. 이전 1시간 데이터에서 세션 active 여부 판단 -> session_id 가 존재하는지
- active된 session 이라면 부여되어있는 session_id 동일하게 부여
- inactive이거나 new session 이라면 새로운 sessio_id 부여


## Airflow dag 개발
Airflow dag 개발 하였습니다. PR을 쪼개지 못한 점 양해부탁드립니다.

### 변경 사항
1. schema 파일 디렉토리 변경하였습니다.
2. qa dag 생성하였습니다
3. Airflow 날짜 및 시간 인자 활용하여 동적 파일 경로로 수정하였습니다.

dodorip committed:
feat: airflow dag 구현

dodorip committed:
refactor: 스키마 파일 디렉토리 변경

dodorip committed:
fix: session_id 생성 로직 변경

- 수정 이유 : 기존 user_id와 session_number 만으로 session_id를 부여할 경우 전체 데이터에선 일관성이 보장되지 않을 가능성 존재
- 수정 내용 : 동일 세션의 start_event_time 컬럼 생성 및 session_id 부여에 활용

Bansng commented:
session_id 컬럼 로직 변경
기존 : user_id & session_number
현재 : session_start_time & user_id & session_number

session_number 정의는?
-> 현재 존재하는 데이터에서 하나의 세션으로 묶이는지 아닌지만 판단
-> user_id 별로 session의 구분
-> 한 시간 데이터 내에서 session을 묶고 거기에 번호를 부여한 것 -> 0,1,2

session_number가 꼭 필요한가요?
-> session_start_time & user_id -> 세션을 식별하는 식별자로서 session_start_time이 그 역할을 하죠

Bansng commented:
https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/dataproc.html#submit-a-job-to-a-cluster
https://airflow.apache.org/docs/apache-airflow-providers-apache-spark/stable/operators.html#sparksubmitoperator

Bansng commented:
airflow 시간변수
https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html

dodorip committed:
fix: session의 active 여부 고려하지 않은 로직 수정

- 이전 30분 데이터 확인하여 active 여부 판단
- active 세션이라면 이미 세션id가 존재하므로 기존 세션id 동일하게 부여
- inactive 세션이라면 새로운 세션id 부여
- session_start_time & user_id로 session_id 식별자 수정
- schema 수정 - session_id 컬럼 추가

dodorip committed:
chore: update .gitignore

dodorip committed:
refactor: session_id 부여 로직 가독성 개선한 코드로 수정

dodorip committed:
fix: pyspark 코드 실행을 위한 SparkSubmitOperator로 수정

dodorip committed:
feat: output csv 생성 추가

dodorip committed:
feat: 파일 경로를 정적 경로에서 동적 경로로 변경

- airflow 실행 날짜 및 시간을 인자로 받아 경로 지정
- 에러 추적을 위해 print 명령어 추가

dodorip committed:
refactor: DAG 파일의 코드 정리

dodorip committed:
feat: airflow sessionization_qa dag 개발

dodorip commented:
ds 및 hour 인자로 file path 동적으로 변경하였습니다.
아래 코드로 과거 데이터 backfill 테스트 까지 완료하였습니다.
airflow dags trigger --exec-date "2019-10-10T14:00:00" hourly_batch_sessionization

start_date 설정 및 backfill 부분까지 완료하였습니다.
